src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -182,15 +187,19 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
                     using var _1 = ArrayBuilder<StateSet>.GetInstance(out var syntaxAnalyzers);
                     using var _2 = ArrayBuilder<StateSet>.GetInstance(out var semanticSpanBasedAnalyzers);
                     using var _3 = ArrayBuilder<StateSet>.GetInstance(out var semanticDocumentBasedAnalyzers);
+                    var seenCompilerAnalyzer = false;
                     foreach (var stateSet in _stateSets)
                     {
                         if (!ShouldIncludeAnalyzer(stateSet.Analyzer, _shouldIncludeDiagnostic, _owner))
```

Comment:
Question about the logic here (and in this type). Presumably diagnostics are cached (so if you call again with the same args the results are just immediately returned. What makes it so that if you ask for something like semantics it knows that syntax hasn't been cached?  Or that if you ask for 'all' that both syntax and semantics are cached?  Thanks!

Reply:
> What makes it so that if you ask for something like semantics it knows that syntax hasn't been cached?

We cache the data per-analyzer by diagnostic kind: https://github.com/dotnet/roslyn/blob/6b02be2614212008d33307e916e37cd53dfb3d52/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs#L335-L347

The data object itself also stores the version information computed by invoking: https://github.com/dotnet/roslyn/blob/6b02be2614212008d33307e916e37cd53dfb3d52/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs#L221-L222

So, when we try to find existing cached data, we fetch the appropriate data object by kind, and then also compare the versions: https://github.com/dotnet/roslyn/blob/6b02be2614212008d33307e916e37cd53dfb3d52/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs#L271-L285 

Note that all the above caching logic that we are talking about is in-proc client side caching. We also cache on the OOP side by holding a weak reference to CompilationWithAnalyzers instance for the active document, so even if there is cache miss in the above in-proc cache, there is good probability of a cache hit on the OOP side and in that case we would incur additional cost for serializing diagnostics back to in-proc but save on not re-executing the analyzers.